// web/src/modules/finance/pages/bank-firm/BankFirmWorkspaceContext.jsx
import { createContext, useCallback, useContext, useEffect, useMemo, useRef, useState } from 'react';
import { useLocation, useSearchParams } from 'react-router-dom';
import { useAuth } from '../../../../auth/AuthContext';
import {
  listFirmBankAccounts,
  createFirmBankAccount,
  updateFirmBankAccount,
  deleteFirmBankAccount,
} from '../../../../services/firmBankAccountService';
import {
  getBankLedger,
  getFirmInternalTxns,
  createFirmBankTransfer,
  createFirmExpenseTxn,
  createFirmInflowTxn,
  updateTxn,
  deleteTxn,
} from '../../services/txnService';

const BankFirmWorkspaceContext = createContext(null);

function filterAccountsByScope(rows, scope) {
  if (scope === 'cash') return rows.filter((a) => a.accountType === 'cash');
  if (scope === 'bank') return rows.filter((a) => a.accountType === 'bank');
  return rows;
}

function readLedgerUrlParams() {
  if (typeof window === 'undefined') {
    return { firm: '', account: '', from: '', to: '' };
  }
  const sp = new URLSearchParams(window.location.search);
  return {
    firm: sp.get('firm') || '',
    account: sp.get('account') || '',
    from: sp.get('from') || '',
    to: sp.get('to') || '',
  };
}

function ledgerUrlParamsMatch(searchParams, { firm, account, from, to }) {
  return (
    (searchParams.get('firm') || '') === firm
    && (searchParams.get('account') || '') === account
    && (searchParams.get('from') || '') === from
    && (searchParams.get('to') || '') === to
  );
}

export function BankFirmWorkspaceProvider({ children }) {
  const { pathname } = useLocation();
  const [searchParams, setSearchParams] = useSearchParams();
  const { hasPermission } = useAuth();
  const forceCashScope = pathname.startsWith('/finance/cash-book');
  const forceBankScope = pathname.startsWith('/finance/bank-reports');
  const cashBookOnly = forceCashScope
    || (hasPermission('cash_book.view') && !hasPermission('invoices.view'));
  const bankOnly = forceBankScope;
  const accountScope = cashBookOnly ? 'cash' : bankOnly ? 'bank' : null;
  const workspaceBase = forceCashScope ? '/finance/cash-book' : '/finance/bank-reports';
  const canSettings = hasPermission('settings.view');
  const canEditOpeningBalance = hasPermission('invoices.edit') || (cashBookOnly && hasPermission('cash_book.edit'));
  const canEditFirmTxn = hasPermission('invoices.edit') || (cashBookOnly && hasPermission('cash_book.edit'));

  const initialLedgerParams = useMemo(() => readLedgerUrlParams(), []);
  const ledgerRestorePendingRef = useRef(
    Boolean(initialLedgerParams.firm && initialLedgerParams.account),
  );

  const [firmCode, setFirmCode] = useState(() => initialLedgerParams.firm);
  const [accounts, setAccounts] = useState([]);
  const [loading, setLoading] = useState(() => Boolean(initialLedgerParams.firm));
  const [msg, setMsg] = useState({ text: '', type: '' });

  const [ledgerAccountId, setLedgerAccountId] = useState(() => initialLedgerParams.account);
  const [ledgerFrom, setLedgerFrom] = useState(() => initialLedgerParams.from);
  const [ledgerTo, setLedgerTo] = useState(() => initialLedgerParams.to);
  const [ledgerRows, setLedgerRows] = useState([]);

  const [xferFrom, setXferFrom] = useState('');
  const [xferTo, setXferTo] = useState('');
  const [xferAmt, setXferAmt] = useState('');
  const [xferDate, setXferDate] = useState(() => new Date().toISOString().slice(0, 10));
  const [xferNote, setXferNote] = useState('');
  const [xferSubmitting, setXferSubmitting] = useState(false);

  const [interFromFirm, setInterFromFirm] = useState('');
  const [interToFirm, setInterToFirm] = useState('');
  const [interFromAccounts, setInterFromAccounts] = useState([]);
  const [interToAccounts, setInterToAccounts] = useState([]);
  const [interFromAcct, setInterFromAcct] = useState('');
  const [interToAcct, setInterToAcct] = useState('');
  const [interAmt, setInterAmt] = useState('');
  const [interDate, setInterDate] = useState(() => new Date().toISOString().slice(0, 10));
  const [interNote, setInterNote] = useState('');

  const [expAcct, setExpAcct] = useState('');
  const [expCat, setExpCat] = useState('other');
  const [expAmt, setExpAmt] = useState('');
  const [expDate, setExpDate] = useState(() => new Date().toISOString().slice(0, 10));
  const [expNote, setExpNote] = useState('');

  const [infAcct, setInfAcct] = useState('');
  const [infCat, setInfCat] = useState('fund_infusion');
  const [infAmt, setInfAmt] = useState('');
  const [infDate, setInfDate] = useState(() => new Date().toISOString().slice(0, 10));
  const [infNote, setInfNote] = useState('');

  const [reportKind, setReportKind] = useState('all');
  const [reportRows, setReportRows] = useState([]);
  const [editTxn, setEditTxn] = useState(null);
  const [editBusy, setEditBusy] = useState(false);
  const [deleteTxnRow, setDeleteTxnRow] = useState(null);
  const [deleteTxnBusy, setDeleteTxnBusy] = useState(false);
  const [deleteRequestReason, setDeleteRequestReason] = useState('');
  const [auditTxn, setAuditTxn] = useState(null);

  const [newName, setNewName] = useState('');
  const [newType, setNewType] = useState('bank');
  const [newOpen, setNewOpen] = useState('0');
  const [newOpenType, setNewOpenType] = useState('debit');
  const [newOpenDate, setNewOpenDate] = useState('');
  const [deleteAccountId, setDeleteAccountId] = useState(null);
  const [deleteAccountBusy, setDeleteAccountBusy] = useState(false);
  const [editAccountId, setEditAccountId] = useState(null);
  const [editOpen, setEditOpen] = useState('');
  const [editOpenType, setEditOpenType] = useState('debit');
  const [editOpenDate, setEditOpenDate] = useState('');
  const [editAccountBusy, setEditAccountBusy] = useState(false);

  const flash = useCallback((text, type = 'info') => {
    setMsg({ text, type });
    setTimeout(() => setMsg({ text: '', type: '' }), 5000);
  }, []);

  const refreshAccounts = useCallback(async ({ silent = false } = {}) => {
    if (!firmCode) {
      setAccounts([]);
      return;
    }
    if (!silent) setLoading(true);
    try {
      const rows = await listFirmBankAccounts(firmCode);
      setAccounts(Array.isArray(rows) ? rows : []);
    } catch (e) {
      if (!silent) {
        flash(e.message || 'Failed to load accounts', 'error');
        setAccounts([]);
      }
    } finally {
      if (!silent) setLoading(false);
    }
  }, [firmCode, flash]);

  useEffect(() => {
    refreshAccounts();
  }, [refreshAccounts]);

  useEffect(() => {
    const nextValues = {
      firm: firmCode,
      account: ledgerAccountId,
      from: ledgerFrom,
      to: ledgerTo,
    };
    if (ledgerUrlParamsMatch(searchParams, nextValues)) return undefined;
    const next = new URLSearchParams(searchParams);
    if (firmCode) next.set('firm', firmCode);
    else next.delete('firm');
    if (ledgerAccountId) next.set('account', ledgerAccountId);
    else next.delete('account');
    if (ledgerFrom) next.set('from', ledgerFrom);
    else next.delete('from');
    if (ledgerTo) next.set('to', ledgerTo);
    else next.delete('to');
    setSearchParams(next, { replace: true });
    return undefined;
  }, [firmCode, ledgerAccountId, ledgerFrom, ledgerTo, searchParams, setSearchParams]);

  const visibleAccounts = useMemo(
    () => filterAccountsByScope(accounts, accountScope),
    [accounts, accountScope],
  );

  const visibleInterFromAccounts = useMemo(
    () => filterAccountsByScope(interFromAccounts, accountScope),
    [interFromAccounts, accountScope],
  );

  const visibleInterToAccounts = useMemo(
    () => filterAccountsByScope(interToAccounts, accountScope),
    [interToAccounts, accountScope],
  );

  useEffect(() => {
    setNewType(accountScope === 'cash' ? 'cash' : 'bank');
  }, [accountScope]);

  useEffect(() => {
    if (!interFromFirm) {
      setInterFromAccounts([]);
      setInterFromAcct('');
      return;
    }
    let cancelled = false;
    listFirmBankAccounts(interFromFirm)
      .then((rows) => {
        if (!cancelled) {
          setInterFromAccounts(Array.isArray(rows) ? rows : []);
          setInterFromAcct('');
        }
      })
      .catch(() => {
        if (!cancelled) {
          setInterFromAccounts([]);
          setInterFromAcct('');
        }
      });
    return () => { cancelled = true; };
  }, [interFromFirm]);

  useEffect(() => {
    if (!interToFirm) {
      setInterToAccounts([]);
      setInterToAcct('');
      return;
    }
    let cancelled = false;
    listFirmBankAccounts(interToFirm)
      .then((rows) => {
        if (!cancelled) {
          setInterToAccounts(Array.isArray(rows) ? rows : []);
          setInterToAcct('');
        }
      })
      .catch(() => {
        if (!cancelled) {
          setInterToAccounts([]);
          setInterToAcct('');
        }
      });
    return () => { cancelled = true; };
  }, [interToFirm]);

  const loadLedger = useCallback(async () => {
    const id = parseInt(ledgerAccountId, 10);
    if (!id) {
      flash('Select a ledger account first', 'error');
      return;
    }
    try {
      const rows = await getBankLedger({ firmBankAccountId: id, dateFrom: ledgerFrom, dateTo: ledgerTo });
      setLedgerRows(rows || []);
    } catch (e) {
      flash(e.message || 'Ledger failed', 'error');
    }
  }, [ledgerAccountId, ledgerFrom, ledgerTo, flash]);

  const onLedgerPage = pathname.endsWith('/ledger');

  useEffect(() => {
    if (!ledgerRestorePendingRef.current || !onLedgerPage) return undefined;
    if (!firmCode || loading) return undefined;
    const id = parseInt(ledgerAccountId, 10);
    if (!id) {
      ledgerRestorePendingRef.current = false;
      return undefined;
    }
    // Account list fetch finished with no accounts — nothing to restore.
    if (accounts.length === 0) {
      ledgerRestorePendingRef.current = false;
      return undefined;
    }
    const exists = visibleAccounts.some((a) => Number(a.id) === id);
    if (!exists) {
      ledgerRestorePendingRef.current = false;
      return undefined;
    }
    ledgerRestorePendingRef.current = false;
    loadLedger();
    return undefined;
  }, [
    onLedgerPage,
    firmCode,
    loading,
    ledgerAccountId,
    accounts.length,
    visibleAccounts,
    loadLedger,
  ]);

  const loadReport = useCallback(async ({ silent = false } = {}) => {
    try {
      const { rows } = await getFirmInternalTxns({ kind: reportKind, perPage: 100 });
      setReportRows(rows || []);
    } catch (e) {
      if (!silent) flash(e.message || 'Report failed', 'error');
    }
  }, [reportKind, flash]);

  const refreshAfterFirmTxn = useCallback(() => {
    void refreshAccounts({ silent: true });
    void loadReport({ silent: true });
  }, [refreshAccounts, loadReport]);

  const submitXfer = useCallback(async (e) => {
    e.preventDefault();
    if (xferSubmitting) return;
    const fromId = parseInt(xferFrom, 10);
    const toId = parseInt(xferTo, 10);
    const fromAcct = accounts.find((a) => Number(a.id) === fromId);
    const toAcct = accounts.find((a) => Number(a.id) === toId);
    const firmFrom = fromAcct?.billingFirmCode || fromAcct?.billing_firm_code;
    const firmTo = toAcct?.billingFirmCode || toAcct?.billing_firm_code;
    if (fromAcct && toAcct && firmFrom && firmTo && firmFrom !== firmTo) {
      flash('Intra transfer requires both accounts in the same billing firm.', 'error');
      return;
    }
    setXferSubmitting(true);
    try {
      await createFirmBankTransfer({
        fromFirmBankAccountId: fromId,
        toFirmBankAccountId: toId,
        amount: parseFloat(xferAmt),
        txnDate: xferDate,
        narration: xferNote,
        transferScope: 'intra',
      });
      flash('Transfer recorded successfully', 'success');
      setXferAmt('');
      setXferNote('');
      refreshAfterFirmTxn();
    } catch (err) {
      flash(err.message || 'Transfer failed', 'error');
    } finally {
      setXferSubmitting(false);
    }
  }, [xferSubmitting, xferFrom, xferTo, xferAmt, xferDate, xferNote, accounts, flash, refreshAfterFirmTxn]);

  const submitInterXfer = useCallback(async (e) => {
    e.preventDefault();
    if (xferSubmitting) return;
    if (interFromFirm === interToFirm) {
      flash('Inter transfer requires different billing firms on each side.', 'error');
      return;
    }
    const fromId = parseInt(interFromAcct, 10);
    const toId = parseInt(interToAcct, 10);
    if (fromId === toId) {
      flash('From and to accounts must differ.', 'error');
      return;
    }
    setXferSubmitting(true);
    try {
      await createFirmBankTransfer({
        fromFirmBankAccountId: fromId,
        toFirmBankAccountId: toId,
        amount: parseFloat(interAmt),
        txnDate: interDate,
        narration: interNote,
        transferScope: 'inter',
      });
      flash('Inter-firm transfer recorded successfully', 'success');
      setInterAmt('');
      setInterNote('');
      refreshAfterFirmTxn();
    } catch (err) {
      flash(err.message || 'Transfer failed', 'error');
    } finally {
      setXferSubmitting(false);
    }
  }, [xferSubmitting, interFromFirm, interToFirm, interFromAcct, interToAcct, interAmt, interDate, interNote, flash, refreshAfterFirmTxn]);

  const submitExp = useCallback(async (e) => {
    e.preventDefault();
    try {
      await createFirmExpenseTxn({
        firmBankAccountId: parseInt(expAcct, 10),
        category: expCat,
        amount: parseFloat(expAmt),
        txnDate: expDate,
        narration: expNote,
      });
      flash('Expense recorded successfully', 'success');
      setExpAmt('');
      setExpNote('');
      refreshAccounts();
      loadReport();
    } catch (err) {
      flash(err.message || 'Expense failed', 'error');
    }
  }, [expAcct, expCat, expAmt, expDate, expNote, flash, refreshAccounts, loadReport]);

  const submitInf = useCallback(async (e) => {
    e.preventDefault();
    try {
      await createFirmInflowTxn({
        firmBankAccountId: parseInt(infAcct, 10),
        category: infCat,
        amount: parseFloat(infAmt),
        txnDate: infDate,
        narration: infNote,
      });
      flash('Inflow recorded successfully', 'success');
      setInfAmt('');
      setInfNote('');
      refreshAccounts();
      loadReport();
    } catch (err) {
      flash(err.message || 'Inflow failed', 'error');
    }
  }, [infAcct, infCat, infAmt, infDate, infNote, flash, refreshAccounts, loadReport]);

  const addAccount = useCallback(async (e) => {
    e.preventDefault();
    if (!canSettings || !firmCode || !newName.trim()) return;
    try {
      await createFirmBankAccount({
        billing_firm_code: firmCode,
        name: newName.trim(),
        account_type: newType,
        opening_balance: Math.abs(parseFloat(newOpen) || 0),
        opening_balance_type: newOpenType,
        opening_balance_date: newOpenDate || null,
      });
      setNewName('');
      setNewOpen('0');
      setNewOpenType('debit');
      flash('Account created', 'success');
      refreshAccounts();
    } catch (err) {
      flash(err.message || 'Create failed', 'error');
    }
  }, [canSettings, firmCode, newName, newType, newOpen, newOpenType, newOpenDate, flash, refreshAccounts]);

  const promptRemoveAccount = useCallback((id) => {
    if (!canSettings) return;
    setDeleteAccountId(id);
  }, [canSettings]);

  const promptEditOpeningBalance = useCallback(
    (account) => {
      if (!canEditOpeningBalance || !account) return;
      setEditAccountId(account.id);
      setEditOpen(String(Math.abs(Number(account.openingBalance) || 0)));
      setEditOpenType(account.openingBalanceType === 'credit' ? 'credit' : 'debit');
      setEditOpenDate(account.openingBalanceDate ? String(account.openingBalanceDate).slice(0, 10) : '');
    },
    [canEditOpeningBalance],
  );

  const closeEditOpeningBalanceModal = useCallback(() => {
    if (!editAccountBusy) {
      setEditAccountId(null);
      setEditOpen('');
      setEditOpenType('debit');
      setEditOpenDate('');
    }
  }, [editAccountBusy]);

  const saveEditOpeningBalance = useCallback(async () => {
    if (!canEditOpeningBalance || editAccountId == null) return;
    setEditAccountBusy(true);
    try {
      await updateFirmBankAccount(editAccountId, {
        opening_balance: Math.abs(parseFloat(editOpen) || 0),
        opening_balance_type: editOpenType,
        opening_balance_date: editOpenDate || null,
      });
      flash('Opening balance updated', 'success');
      setEditAccountId(null);
      setEditOpen('');
      setEditOpenType('debit');
      setEditOpenDate('');
      refreshAccounts();
    } catch (err) {
      flash(err.message || 'Update failed', 'error');
    } finally {
      setEditAccountBusy(false);
    }
  }, [canEditOpeningBalance, editAccountId, editOpen, editOpenType, editOpenDate, flash, refreshAccounts]);

  const confirmRemoveBankAccount = useCallback(async () => {
    if (!canSettings || deleteAccountId == null) return;
    setDeleteAccountBusy(true);
    try {
      await deleteFirmBankAccount(deleteAccountId);
      flash('Account deleted', 'success');
      setDeleteAccountId(null);
      refreshAccounts();
    } catch (err) {
      flash(err.message || 'Delete failed', 'error');
    } finally {
      setDeleteAccountBusy(false);
    }
  }, [canSettings, deleteAccountId, flash, refreshAccounts]);

  const closeDeleteModal = useCallback(() => {
    if (!deleteAccountBusy) setDeleteAccountId(null);
  }, [deleteAccountBusy]);

  const openEditFirmTxn = useCallback((row) => {
    if (!canEditFirmTxn || !row) return;
    setEditTxn({ ...row });
  }, [canEditFirmTxn]);

  const closeEditFirmTxn = useCallback(() => {
    if (!editBusy) setEditTxn(null);
  }, [editBusy]);

  const saveEditFirmTxn = useCallback(
    async (form) => {
      if (!editTxn?.id || !canEditFirmTxn) return;
      const reason = String(form.requestReason || '').trim();
      if (!reason) {
        flash('Please enter a reason for this change.', 'error');
        return;
      }
      setEditBusy(true);
      try {
        const type = editTxn.txnType;
        let payload = {};
        if (type === 'firm_bank_transfer') {
          const scope = form.transferScope || editTxn.transferScope || 'intra';
          payload = {
            txn_date: form.txnDate,
            amount: parseFloat(form.amount),
            narration: form.narration,
            from_firm_bank_account_id: parseInt(form.fromAccountId, 10),
            to_firm_bank_account_id: parseInt(form.toAccountId, 10),
            transfer_scope: scope,
            request_reason: reason,
          };
        } else {
          payload = {
            txn_date: form.txnDate,
            amount: parseFloat(form.amount),
            narration: form.narration,
            firm_bank_account_id: parseInt(form.bankAccountId, 10),
            firm_expense_category: form.category,
            request_reason: reason,
          };
        }
        const result = await updateTxn(editTxn.id, payload);
        if (result?.pendingLedgerChange) {
          flash(result.queuedMessage || 'Change submitted for Super Admin approval.', 'success');
        } else {
          flash('Transaction updated.', 'success');
        }
        setEditTxn(null);
        loadReport();
        refreshAccounts();
      } catch (err) {
        flash(err.message || 'Update failed', 'error');
      } finally {
        setEditBusy(false);
      }
    },
    [editTxn, canEditFirmTxn, flash, loadReport, refreshAccounts],
  );

  const promptDeleteFirmTxn = useCallback(
    (row) => {
      if (!canEditFirmTxn || !row) return;
      setDeleteTxnRow(row);
    },
    [canEditFirmTxn],
  );

  const closeDeleteFirmTxnModal = useCallback(() => {
    if (!deleteTxnBusy) {
      setDeleteTxnRow(null);
      setDeleteRequestReason('');
    }
  }, [deleteTxnBusy]);

  const confirmDeleteFirmTxn = useCallback(async () => {
    if (!deleteTxnRow?.id || !canEditFirmTxn) return;
    const reason = deleteRequestReason.trim();
    if (!reason) {
      flash('Please enter a reason for this cancellation.', 'error');
      return;
    }
    setDeleteTxnBusy(true);
    try {
      const result = await deleteTxn(deleteTxnRow.id, { request_reason: reason });
      if (result?.pendingLedgerChange) {
        flash(result.queuedMessage || 'Cancel submitted for Super Admin approval.', 'success');
      } else {
        flash('Transaction cancelled.', 'success');
      }
      setDeleteTxnRow(null);
      loadReport();
      refreshAccounts();
    } catch (err) {
      flash(err.message || 'Delete failed', 'error');
    } finally {
      setDeleteTxnBusy(false);
    }
  }, [deleteTxnRow, deleteRequestReason, canEditFirmTxn, flash, loadReport, refreshAccounts]);

  const openAuditFirmTxn = useCallback((row) => {
    if (row?.id) setAuditTxn(row);
  }, []);

  const closeAuditFirmTxn = useCallback(() => setAuditTxn(null), []);

  const value = useMemo(
    () => ({
      cashBookOnly,
      bankOnly,
      workspaceBase,
      canSettings,
      canEditOpeningBalance,
      canEditFirmTxn,
      firmCode,
      setFirmCode,
      accounts: visibleAccounts,
      loading,
      msg,
      flash,
      ledgerAccountId,
      setLedgerAccountId,
      ledgerFrom,
      setLedgerFrom,
      ledgerTo,
      setLedgerTo,
      ledgerRows,
      loadLedger,
      xferFrom,
      setXferFrom,
      xferTo,
      setXferTo,
      xferAmt,
      setXferAmt,
      xferDate,
      setXferDate,
      xferNote,
      setXferNote,
      xferSubmitting,
      submitXfer,
      interFromFirm,
      setInterFromFirm,
      interToFirm,
      setInterToFirm,
      interFromAccounts: visibleInterFromAccounts,
      interToAccounts: visibleInterToAccounts,
      interFromAcct,
      setInterFromAcct,
      interToAcct,
      setInterToAcct,
      interAmt,
      setInterAmt,
      interDate,
      setInterDate,
      interNote,
      setInterNote,
      submitInterXfer,
      expAcct,
      setExpAcct,
      expCat,
      setExpCat,
      expAmt,
      setExpAmt,
      expDate,
      setExpDate,
      expNote,
      setExpNote,
      submitExp,
      infAcct,
      setInfAcct,
      infCat,
      setInfCat,
      infAmt,
      setInfAmt,
      infDate,
      setInfDate,
      infNote,
      setInfNote,
      submitInf,
      reportKind,
      setReportKind,
      reportRows,
      loadReport,
      editTxn,
      editBusy,
      openEditFirmTxn,
      closeEditFirmTxn,
      saveEditFirmTxn,
      deleteTxnRow,
      deleteTxnBusy,
      deleteRequestReason,
      setDeleteRequestReason,
      promptDeleteFirmTxn,
      closeDeleteFirmTxnModal,
      confirmDeleteFirmTxn,
      auditTxn,
      openAuditFirmTxn,
      closeAuditFirmTxn,
      newName,
      setNewName,
      newType,
      setNewType,
      newOpen,
      setNewOpen,
      newOpenType,
      setNewOpenType,
      newOpenDate,
      setNewOpenDate,
      addAccount,
      promptRemoveAccount,
      deleteAccountId,
      deleteAccountBusy,
      confirmRemoveBankAccount,
      closeDeleteModal,
      editAccountId,
      editOpen,
      setEditOpen,
      editOpenType,
      setEditOpenType,
      editOpenDate,
      setEditOpenDate,
      editAccountBusy,
      promptEditOpeningBalance,
      closeEditOpeningBalanceModal,
      saveEditOpeningBalance,
    }),
    [
      cashBookOnly,
      bankOnly,
      workspaceBase,
      canSettings,
      canEditOpeningBalance,
      canEditFirmTxn,
      firmCode,
      visibleAccounts,
      visibleInterFromAccounts,
      visibleInterToAccounts,
      loading,
      msg,
      flash,
      ledgerAccountId,
      ledgerFrom,
      ledgerTo,
      ledgerRows,
      loadLedger,
      xferFrom,
      xferTo,
      xferAmt,
      xferDate,
      xferNote,
      xferSubmitting,
      submitXfer,
      interFromFirm,
      interToFirm,
      interFromAcct,
      interToAcct,
      interAmt,
      interDate,
      interNote,
      submitInterXfer,
      expAcct,
      expCat,
      expAmt,
      expDate,
      expNote,
      submitExp,
      infAcct,
      infCat,
      infAmt,
      infDate,
      infNote,
      submitInf,
      reportKind,
      reportRows,
      loadReport,
      editTxn,
      editBusy,
      openEditFirmTxn,
      closeEditFirmTxn,
      saveEditFirmTxn,
      deleteTxnRow,
      deleteTxnBusy,
      deleteRequestReason,
      setDeleteRequestReason,
      promptDeleteFirmTxn,
      closeDeleteFirmTxnModal,
      confirmDeleteFirmTxn,
      auditTxn,
      openAuditFirmTxn,
      closeAuditFirmTxn,
      newName,
      newType,
      newOpen,
      newOpenType,
      newOpenDate,
      addAccount,
      promptRemoveAccount,
      deleteAccountId,
      deleteAccountBusy,
      confirmRemoveBankAccount,
      closeDeleteModal,
      editAccountId,
      editOpen,
      editOpenType,
      editOpenDate,
      editAccountBusy,
      promptEditOpeningBalance,
      closeEditOpeningBalanceModal,
      saveEditOpeningBalance,
    ],
  );

  return <BankFirmWorkspaceContext.Provider value={value}>{children}</BankFirmWorkspaceContext.Provider>;
}

export function useBankFirmWorkspace() {
  const ctx = useContext(BankFirmWorkspaceContext);
  if (!ctx) throw new Error('useBankFirmWorkspace must be used within BankFirmWorkspaceProvider');
  return ctx;
}
