// web/src/auth/ProtectedRoute.jsx
import { Navigate } from 'react-router-dom';
import { useAuth } from './AuthContext';

/**
 * Wraps private routes.
 *
 * - Redirects to /login when the user is not authenticated.
 * - Shows a 403 "Access Denied" screen when `requiredPermission` is set
 *   and the current user lacks that permission.
 * - Renders a blank screen while the token validation is in progress.
 *
 * `associateOnly` — only users with role associate.
 * `partnerOnly`   — only users with role partner.
 * `staffOnly` — blocks associate/partner/client roles from staff UI.
 */
export default function ProtectedRoute({
  children,
  requiredPermission,
  requiredAnyPermissions,
  associateOnly,
  partnerOnly,
  clientOnly,
  staffOnly,
}) {
  const { isAuthenticated, loading, hasPermission, hasAnyPermission, user } = useAuth();

  // Wait for the token validation round-trip before deciding
  if (loading) {
    return (
      <div style={{
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        minHeight: '100vh',
        background: 'var(--portal-bg, #f8fafc)',
      }}>
        <div style={{ color: '#64748b', fontSize: 14 }}>Loading…</div>
      </div>
    );
  }

  if (!isAuthenticated) {
    const savedPortal = sessionStorage.getItem('login_portal');
    const loginTo = savedPortal ? `/login?portal=${savedPortal}` : '/login';
    return <Navigate to={loginTo} replace />;
  }

  const role = user?.role || '';

  if (staffOnly && role === 'associate') {
    return <Navigate to="/associate" replace />;
  }

  if (staffOnly && role === 'partner') {
    return <Navigate to="/partner" replace />;
  }

  if (staffOnly && role === 'client') {
    return <Navigate to="/client" replace />;
  }

  if (associateOnly && role !== 'associate') {
    return <Navigate to="/" replace />;
  }

  if (partnerOnly && role !== 'partner') {
    return <Navigate to="/" replace />;
  }

  if (clientOnly && role !== 'client') {
    return <Navigate to="/" replace />;
  }

  if (requiredAnyPermissions?.length && !hasAnyPermission(requiredAnyPermissions)) {
    return (
      <div style={{
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
        justifyContent: 'center',
        minHeight: '100vh',
        background: 'var(--portal-bg)',
        gap: 16,
      }}>
        <div style={{ fontSize: 64 }}>🚫</div>
        <h1 style={{ margin: 0, color: '#1e293b', fontSize: 28, fontWeight: 700 }}>
          403 — Access Denied
        </h1>
        <p style={{ color: '#64748b', margin: 0 }}>
          You don&apos;t have permission to view this page.
        </p>
      </div>
    );
  }

  if (requiredPermission && !hasPermission(requiredPermission)) {
    return (
      <div style={{
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
        justifyContent: 'center',
        minHeight: '100vh',
        background: 'var(--portal-bg)',
        gap: 16,
      }}>
        <div style={{ fontSize: 64 }}>🚫</div>
        <h1 style={{ margin: 0, color: '#1e293b', fontSize: 28, fontWeight: 700 }}>
          403 — Access Denied
        </h1>
        <p style={{ color: '#64748b', margin: 0 }}>
          You don&apos;t have permission to view this page.
        </p>
      </div>
    );
  }

  return children;
}
